### ISSUE TITLE
Improve farming

### ISSUE THREAD

**Issue by MarkuBu** [D00Med/farlands#140]
Farlands adds some new items to farming, but the only food I use is bread. There are two main reasons:

1. Bread gives the most saturation as far as I could see
2. Crops like potatoes rot too fast, as mentioned in #127 

To make farming more attractive some thoughts for discussion.

- Increase growing time of the crops and let them stay much longer in "mature" state.
- Decrease saturation of pure bread and add some food recipes like a chicken sandwich, potato soup
- Add some more crops like lettuce, soy and so on.
- Add salt to the map
- Some crops only can be found in backyards or can be purchased as seeds from a trader.
- Currently it is faster if you harvest with a sword. Would make more sense with a scythe or a sickle. (Ironically the german word for scythe is "Sense" ;) )
- Add dough to make pie

Some more will follow

**Comment by tobyplowy** [comment]
> Increase growing time of the crops and let them stay much longer in "mature" state.

Yeh that was a problem d00med tryed to fix it but it didn't work :/

> Decrease saturation of pure bread and add some food recipes like a chicken sandwich, potato soup

Yes good idea there really should be more crafteble food

> Add some more crops like lettuce, soy and so on.

That was the plan :)

> Add salt to the map

For food crafting yes?

> Some crops only can be found in backyards or can be purchased as seeds from a trader.

Good idea also monsters can drop some things too

> Currently it is faster if you harvest with a sword. Would make more sense with a scythe or a sickle. 

True :)

> (Ironically the german word for scythe is "Sense" ;) )

Huh interesting :)

> Add dough to make pie

Yes!

Now only one problem I can only make models and textures I cat program 
He's the programmer on board but he has no time so yeh

> What I really want to do is to improve farming. 

Oh well that's convenient XD
Are you ok with working together with me?
I do art you program?

**Comment by MarkuBu** [comment]
Ok, let me take a look to the code... Tomorrow

**Comment by tobyplowy** [comment]
@MarkuBu Ok :)

**Comment by MarkuBu** [comment]
> Yeh that was a problem d00med tryed to fix it but it didn't work :/

It's a bit confusing what happens there. The general problem is, that the farming mod doesn't know a "mature" or "rot" state. It only knows states and it automatically steps from the first to the last state.

I try to add a mature state and stay in this state for a while.

![screenshot_20171011_105828](https://user-images.githubusercontent.com/16163058/31431222-5131160c-ae73-11e7-8fc6-b28486cc6b74.png)

The potatoes have a mature state and a 5% chance to rot, the corn has no mature state and grows as before. 

But another issue is, that wheat has 7 states until mature, cotton has 8, but potatoes, corn and so on only have 4 states until mature. The number of stages defines the growing time. 

I suggest to add more states to seasonal crops like potatoes and corn, crops like lettuce or radish can have 3 or 4 states because they grow faster.

And I would increase the grow interval because I think, crops grow too fast

**Comment by MarkuBu** [comment]
And by the way... radish is spelled with one "d", not with two

**Comment by tobyplowy** [comment]
@MarkuBu is it possible to make it take longer to grow with code?
Making these stage textures is a nightmare it takes forever :/
I'm willing to do it though.

**Comment by tobyplowy** [comment]
> And by the way... radish is spelled with one "d", not with two

Yeah you're probably going to find the couple throughout farlands XD

**Comment by MarkuBu** [comment]
Currently the average grow time from stage to stage is the same for all corps. But I think it is possible to add individual grow times for each crop, independent of the number of stages. Probably I can use ingame days as value. Just say, "potatoes need 10 days to grow, radish 3 days". 

And we should add a rot state for all crops anyway, even for wheat and cotton, to make it consistent.

**Comment by MarkuBu** [comment]
This runs now for about 1.5 hours. I doubled the average grow time and the 5% chance to rot. Looks better for me. You don't need to maintain the fields continuously

![screenshot_20171011_124447](https://user-images.githubusercontent.com/16163058/31436100-15322d76-ae82-11e7-9a31-ae87c287c393.png)

**Comment by tobyplowy** [comment]
@MarkuBu looking good :)

**Comment by tobyplowy** [comment]
> potatoes need 10 days to grow, radish 3 days

Thats Still abit randomized right so they Don't all grow at the same Time ;)

**Comment by MarkuBu** [comment]
Of course. Currently you have an average grow time per state in seconds. And each state has a random range based on that time.

I want to calculate this average grow time per state based on the seconds per day and the number of states. With this method the number of states or the length of a day doesn't matter and we have reasonable grow times because I really think, everything grows much too fast right now

**Comment by tobyplowy** [comment]
@MarkuBu ok good
 also how do you want me to give you the textures do I upload them in this issue and you add them to your branch
(I prefer to do it this way)

Or do you have another idea

**Comment by MarkuBu** [comment]
You can upload wherever you want

But I want to reduce the randomization a bit. The difference between the states can be to high. I have cotton in state 8 and in state 3. It's a little too much

**Comment by tobyplowy** [comment]
> But I want to reduce the randomization a bit. The difference between the states can be to high. I have cotton in state 8 and in state 3. It's a little too much

Jeh that is pretty extreme So jeh go ahead

**Comment by MarkuBu** [comment]
I've updated the first comment. I removed things that are not part of the farming mod. Additional recipes should be added with a new mod.

The farming mod should only contain stuff that is related to farming which is soil, tools, crops and the growing of the crops.

I will add some suggestions for new crops to the first post

**Comment by tobyplowy** [comment]
> I've updated the first comment 

i cant find it

**Comment by MarkuBu** [comment]
I mean the first comment in this issue

https://github.com/D00Med/farlands/issues/140#issue-264209856

**Comment by MarkuBu** [comment]
The code from the farming mod confuses me. I have to understand it first, before I can make major changes

**Comment by tobyplowy** [comment]
@MarkuBu you can rewerite the whole farming mod if youd like (make sure it's compatible with V1 world saves :)

**Comment by MarkuBu** [comment]
Pushed some minor changes to 1.1

Grow time is longer. 
Crops stay much longer in mature state
Grow speed is less (but still) random

**Comment by tobyplowy** [comment]
@MarkuBu can you inprove the watering can (so it can be used as a farming tool?)

I whant it to have a durability bar and When The durability bar is empty you need to refill it with water 
Also it's not possible to water crops for some reason it's only possible to make saplings grow

D00Med made the original watering can

**Comment by tobyplowy** [comment]
> Pushed some minor farming changes to 1.1
> 
> Grow time is longer.
> Crops stay much longer in mature state
> Grow speed is less (but still) random

Amazing thanks again for all your work and time it really means alot to me :)

**Comment by MarkuBu** [comment]
Try it first before you cheer ;-)

**Comment by tobyplowy** [comment]
@MarkuBu I'm testing now, but you still deserve a thank you for your time and work you know :)

**Comment by tobyplowy** [comment]
@MarkuBu I found that Wheat and Cotton grow verry slow can it be faster

**Comment by tobyplowy** [comment]
@MarkuBu I'm also inproveing the crop grow stages

So it's easier to see when they are ready to be harvested

**Comment by MarkuBu** [comment]
It grows slow because I just increased the stage time for now, but it still has 7/8 stages instead of 4. It's a bit more work to make it more flexible.

But nevertheless, I think longer times for food production would be better. I don't know what I should do with all the eggs and chicken ;-)

![screenshot_20171019_213100](https://user-images.githubusercontent.com/16163058/31790253-1f86b1f4-b515-11e7-9b5c-646efc2ecf05.png)

**Comment by tobyplowy** [comment]
@MarkuBu rotten cotton (cotton dead)
![rotten cotton](https://user-images.githubusercontent.com/16853304/31790408-a904dadc-b515-11e7-92ad-dada779bee0b.png)

also thats alot of eggs i wish this was real i love eggs :/

**Comment by tobyplowy** [comment]
@MarkuBu can you add temp textures for rotten wheat and the new crops you sugested 
(just add a solid color 16X16 image as texture) so i can work on it later and its alredy in the game

**Comment by tobyplowy** [comment]
rotten_wheat
![rotten_wheat](https://user-images.githubusercontent.com/16853304/31790757-d806f936-b516-11e7-9ac1-8cff2ceceaef.png)

**Comment by tobyplowy** [comment]
@MarkuBu I just tested the crop grow times are perfect :)
Wheat and Cotton need to go a bit faster tho also rotting needs to go a little bit faster 

Nice job

**Comment by tobyplowy** [comment]
@MarkuBu are you still sick?

**Comment by MarkuBu** [comment]
Ok, I think I will rewrite the farming mod. Using farming redo as base is a mess. Probably I will use default farming mod as base and go from there.

And I will split farming and food recipes into two mods

**Comment by MarkuBu** [comment]
Some suggestions. 

- You can't find any crop in the wild. To get some progress to the game most crops only can be found in backyards, in chests, can be purchased by traders or dropped e.g. by mobs. 
- Crops can grow on dry soil, on wet soil and on fertile soil (compost). On dry soil they grow slow and rot fast, but you can grow something even if you have no water or bucket. On fertile soil they grow fast and don't rot. 
- In cold biomes crops grow slower
- Crops should support grow speedup with bone meal or similar
- Grow speed should depend on light level, soil, temperature...

I don't like the idea of to much maintenance like watering the soil with a watering can.

**Comment by tobyplowy** [comment]
> Ok, I think I will rewrite the farming mod.

I'm excited :)

> You can't find any crop in the wild.

Eeeemmmm yes you can at least in v1 not sure about 1.1 tho :/

> Crops can grow on dry soil, on wet soil and on fertile soil (compost). On dry soil they grow slow and rot fast, but you can grow something even if you have no water or bucket. On fertile soil they grow fast and don't rot.

I agree with the above but if the soil is not wet it should grow really slow 
(Like painfully slow becouse it's not hard to get water)

Can you change dry soil into wet soil if it's raining without having alot of performance issues?

> In cold biomes crops grow slower

Oh man I really like that idea 

> Crops should support grow speedup with bone meal or similar

Yes :)
Can the watering can be used? (Idea below)

> Grow speed should depend on light level, soil, temperature...

If this doesn't make too much lag than yes also how can minetest know what temperature it is?
Maybe we can add air blocks for the different biomes (this will also allow us to change the sky color)
2 good things in one :)

> Use 'on_flood' to drop crops when flooded with water

Yes we should have added that from the beginning :/

> I don't like the idea of to much maintenance

Well the fun of farming is that it takes some time but we can settle an agreement 
I whant more work for farming you whant less but im very cooperative unlike other people 
(minetest dev team)

> like watering the soil with a watering can.

Maybe it can be used instead of bone meal it has the same functions but you only need to refill it every now and then good idea yes or no?

**Comment by MarkuBu** [comment]
> Eeeemmmm yes you can at least in v1 not sure about 1.1 tho :/

I know that it is currently possible, but for future versions this should be removed. It won't break old worlds. You just wont find crops in newly explored areas.

A game needs some progress. Walking through a grassland and find 9 of 10 crops is not very satisfying. You need some progress. For example: find wheat seed, grow wheat, get bread. Find a house with some potatoes in a chest, grow more potatoes, make baked potatoes. Find some coins in a dungeon, go to a trader an buy some whatever seeds. Grow whatever to make a delicious meal with potatoes, meat and whatever .... 

> Can the watering can be used? (Idea below)

Well, if you have to use some kind of fertilizer (e.g. compost tea) instead of pure water because pure water would make it to easy to grow something fast.

> I whant more work for farming you whant less

If farming takes too much time you don't have time for other things like building or traveling. Especially traveling becomes nearly impossible because you have to return home every time you need to maintain your farm.

The watering can in combination with some kind of fertilizer can be used to grow crops faster (like with bone meal), but not for watering the soil to keep it wet.

**Comment by tobyplowy** [comment]
> whatever

I really like whatever it's my favorite food XD
But all joking aside I agree

> The watering can in combination with some kind of fertilizer can be used to grow crops faster (like with bone meal), but not for watering the soil to keep it wet.

Yes ok

Is it possible to make the crops grow at random if you use fertilizer
So you need to fertilise it more Than ones to get a full crop?

**Comment by MarkuBu** [comment]
> Is it possible to make the crops grow at random if you use fertilizer
So you need to fertilise it more Than ones to get a full crop?

I think about to add an API to make it easy to add fertilizer with different chances

But something else: I don't like the fact that it is enough to have a water source or floating water within 3 nodes, even if it is 3 nodes below or flying 3 nodes above the crops. Water should be at the same level as the soil

![screenshot_20171028_164550](https://user-images.githubusercontent.com/16163058/32136073-63aa95e4-bc09-11e7-849a-bad3e5de756c.png)

![screenshot_20171028_175844](https://user-images.githubusercontent.com/16163058/32136090-b06a5d88-bc09-11e7-8c0d-1f71e6985fc3.png)

**Comment by tobyplowy** [comment]
> I think about to add an API to make it easy to add fertilizer with different chances

Nice :)

> But something else: I don't like the fact that it is enough to have a water source or floating water within 3 nodes, even if it is 3 nodes below or flying 3 nodes above the crops. Water should be at the same level as the soil

I never even knew that was possible but yes I agree 
Also can you increase the distance of water effect to 4 blocks far instead of 3?

**Comment by MarkuBu** [comment]
> Also can you increase the distance of water effect to 4 blocks far instead of 3?

Already done

![screenshot_20171028_185425](https://user-images.githubusercontent.com/16163058/32136639-7ed7d39c-bc11-11e7-8e99-44f62287d304.png)

Now the water source has to be at the same level as the soil and you can water a 9x9 field instead of 7x7

But don't be confused. This is not farlands. Currently I use default MTG. I just modify the default farming mod

**Comment by tobyplowy** [comment]
@MarkuBu it will be added to farlands later right?

Also do you think pigs shuld be added to farlands?

**Comment by MarkuBu** [comment]
We should open a new issue for animals

**Comment by D00Med** [comment]
@MarkuBu Something good about anything like farming being more complicated, is that it extends the depth of the game. If you look at the other aspects of Farlands, complexity is sort of a running theme. 
So here are my thoughts:
Farming could be complex, but it should tie in with the village building aspect. When I add npc vocations, we can have farmers that grow crops in nearby areas, without it being so complicated. But if the player wants to grow their own, then it is harder. So it's easy only if players have put work into their village/s.
I like the idea that crops shouldn't grow in the wild, but maybe instead it would be better for them to just be rare? And farmer npc's could sell seeds (maybe just make the traders do this for now).
And if crops can grow in dry soil, I suggest that they should grow slowly, but spoil quickly.

**Comment by tobyplowy** [comment]
@D00Med i agree :)

**Comment by MarkuBu** [comment]
Well, I forgot something important. Nothing grows if no player is near the plants. So you can walk around as long as you want.

But if you are building near your farm plants are growing.

And it is a difference between complexity and annoyance. And it is annoying if you have to interrupt whatever you do every other minute to maintain your farm.

**Comment by MarkuBu** [comment]
just a few ideas for more complexity. 
Wheat
If you harvest you only get wheat, but no seeds. You need a scyth to harvest wheat. To get seeds you need a flail and you get straw and seeds. If you do it by hand you get less seeds. Wheat is attacked by crows. You need to build a scarecrow to protect your fields. To build a scarecrow you need to find a pumpkin and maybe some leather or cotton fabric.

Cotton
Cotton needs a lot of water. Instead of growing just on wet farming soil each plant needs water nearby.

To prevent pest some kind of mixed culture can be used. For example Carrots can be infested by white flies, but onions as neighbors can help.

Just adding repetitive tasks don't add complexity, they are just annoying. You need to get some progress to farming. 

Next step could be some kind of harvester and planter or, if it is more important for you, villager who work at a farm. But this is getting more complex

**Comment by D00Med** [comment]
@MarkuBu 
Those are good ideas. Crows would be cool, @tobyplowy could you make a model for those?
Would it be good to have pipelines for water? If I added that, it would be a good use for machines.

**Comment by tobyplowy** [comment]
> could you make a model for those?

Yes I can but maybe we can use the bird model that we already have?

**Comment by tobyplowy** [comment]
> Would it be good to have pipelines for water? If I added that, it would be a good use for machines.

Yes if sprinklers get added

**Comment by D00Med** [comment]
@tobyplowy It's not really crow-shaped. 
Sprinklers are a good idea :)

**Comment by tobyplowy** [comment]
> It's not really crow-shaped.

For quality reasons I will make a new model :)
Added to my to-do list

**Comment by D00Med** [comment]
@tobyplowy Thankyou :)

**Comment by MarkuBu** [comment]
How do we get seeds? That is the next question to which I would like an answer.

The default farming mod assumes that harvesting always drops a crop and a seed. But that doesn't makes sense. 

My suggestions:
- Wheat: If you harvest wheat you get wheat. From that wheat you get straw and seeds. From the seeds you can make flour or you can plant them.
- Potatoes: If you harvest you get potatoes. The potatoes can be planted as seeds as usual
- Cotton: If you harvest you get cotton balls. The seeds are inside the cotton balls. If you separate them you get cotton and seeds.
- Pumpkin, Melon: If you harvest pumpkins/melons you don't get "slices". You get a pumpkin or melon. You can cut them in slices an get the seeds out of the "meat".

Now it gets more complicated:
- Carrots: in nature carrots need to flower to get seeds, but they flower in the second year. 
- Same with lettuce. You can't plant a lettuce plant to grow new lettuce

The question: Should salad, carrots, etc. stay as it is? The crop is also the seed?

**Comment by MarkuBu** [comment]
The thing is, that the default farming mod from MTG defines crops and seeds for all plants. If you harvest you always get a crop and a seed.

We need to define now how farming should look like in Farlands. 
- Crops and seeds for all plants like in MTG?
- Crops can be seeds, even if it is not realistic like carrots in farming redo?
- A more realistic approach where some crops can be seeds, some crops need to flower...?
- Something else?

**Comment by tobyplowy** [comment]
> Crops can be seeds, even if it is not realistic like carrots in farming redo?

Yes that one

> Pumpkin, Melon: If you harvest pumpkins/melons you don't get "slices". You get a pumpkin or melon. You can cut them in slices an get the seeds out of the "meat".

Yes that's a better idea

NOTE we whant to make farlands complicated but a kid shuld still be abel to understand it :)

I dount like the flowering idea

**Comment by MarkuBu** [comment]
Ok, I agree with that too. Just want to clarify. 

The thing is: default MTG farming simplifies everything to a common ground. All crops are handled identical. farming redo is a bit messy because you have to define each state for each crop. But both have the problem, that they define a common growing routine. Too many if's and then's to decide what should happen. 

Let's define farming:
- Each crop has a grow time, independent from the number of grow states.
- Each crop has a mature state, where it stays for a while.
- Each crop has a rot state after mature, which is the last state.
- Crops like potatoes or carrots don't drop seeds because the crop itself is the seed.
- Pumpkins and melons are harvested in one peace and can be crafted into slices. Slices can be used as seeds.
- Wheat and cotton drop seeds and the crop when harvested as it is now.

And one important thing: farming should only contain seasonal crops and not "perennials" like fruits and berry bushes

**Comment by tobyplowy** [comment]
> Each crop has a grow time, independent from the number of grow states.

Yes

> Each crop has a mature state, where it stays for a while.

Yes

> Each crop has a rot state after mature, which is the last state.

I'm repeating myself but yes

Maybe rotten crops can be used as compost?

> Crops like potatoes or carrots don't drop seeds because the crop itself is the seed.

Correct

> Pumpkins and melons are harvested in one peace and can be crafted into slices. Slices can be used as seeds.

Maybe slices need to be crafted in to seeds first? 

> Wheat and cotton drop seeds and the crop when harvested as it is now

Yes

> And one important thing: farming should only contain seasonal crops and not "perennials" like fruits and berry bushes

Why not?

**Comment by MarkuBu** [comment]
> Maybe rotten crops can be used as compost?

Of course, but mostly anything from plants can be used to make compost

> Maybe slices need to be crafted in to seeds first?

Maybe, but not necessarily. 

> Why not?

You already have fruit trees and berry bushes. You add fruits to the default trees and bushes. I would extend this as I mentioned in #153. You can make fruit trees from sapling + fruit and maybe berry bushes from bush sapling + berry. But they are different kind of plants. Everything that grows on trees or bushes should be handled as tree or bush. 

Other plants like strawberries or rhubarb are perennials, but I would treat them like seasonal crops

**Comment by tobyplowy** [comment]
@markubu ok

**Comment by tobyplowy** [comment]
@MarkuBu @D00Med crow is done
[crow.zip](https://github.com/D00Med/farlands/files/1431804/crow.zip)
![crow pic](https://user-images.githubusercontent.com/16853304/32240980-2452fe00-be6f-11e7-807f-67e4d5f1576e.png)

**Comment by D00Med** [comment]
@tobyplowy Awesome, it looks great!

**Comment by MarkuBu** [comment]
Nice job

**Comment by tobyplowy** [comment]
@MarkuBu @D00Med thanks :)

**Comment by tobyplowy** [comment]
@MarkuBu @D00Med shuld we add slimes aswell (they should also eat crops)

**Comment by MarkuBu** [comment]
> shuld we add slimes aswell

Hmm, no. Not now

**Comment by tobyplowy** [comment]
@MarkuBu shuld we add moth traps someting like this
https://starbounder.org/Moth_Trap

They give silk and that can be used for stuff like making string and so on

**Comment by MarkuBu** [comment]
Interesting idea, but shouldn't added to farming.

I still have no good idea how to implement farming in a nice way.

**Comment by tobyplowy** [comment]
@MarkuBu this was actually in my original plan for farming when me and d00med started farlands
Can you add it please?

**Comment by tobyplowy** [comment]
@MarkuBu I already have a model

**Comment by tobyplowy** [comment]
@MarkuBu 
![moth trap pic](https://user-images.githubusercontent.com/16853304/32520172-5e1db49a-c40f-11e7-9693-94ca70c54c0d.png)

**Comment by tobyplowy** [comment]
^ moth trap ^
